ansible_base/lib/routers/association_resource_router.py

```diff
@@ -120,45 +148,82 @@ def disassociate(self, request, **kwargs):
 
         return Response(status=status.HTTP_204_NO_CONTENT)
 
-    def get_serializer_class(self):
+    def get_association_queryset(self) -> QuerySet:
+        """Return queryset for instances field of the serializer used for association"""
         if self.queryset:
-            qs = self.queryset
-            cls = self.queryset.model
-        else:
-            cls = self.serializer_class.Meta.model
-            qs = cls.objects.all()
-
-        if self.action == 'disassociate':
-            return basic_association_serializer_factory(qs)
-        elif self.action == 'associate':
-            if 'ansible_base.rbac' in settings.INSTALLED_APPS and permission_registry.is_registered(cls):
-                return filtered_association_serializer_factory(cls, qs)
-            elif 'ansible_base.rbac' in settings.INSTALLED_APPS and cls._meta.model_name == 'user':
-                return UserAssociationSerializer
-            else:
-                return basic_association_serializer_factory(qs)
+            return self.queryset
+        cls = self.serializer_class.Meta.model
+        return cls.objects.all()
+
+    def get_serializer_class(self):
+        if self.action in ('disassociate', 'associate'):
+            rel_name = self.association_fk.replace('_', ' ').title().replace(' ', '')
+            cls_name = f'{self.parent_viewset.__name__}{rel_name}{self.action.title()}Serializer'
+
+            if cls_name not in serializer_registry:
+                qs = self.get_association_queryset()
+                if self.action == 'associate':
+                    serializer_registry[cls_name] = type(cls_name, (FilteredAssociationSerializer,), {'target_queryset': qs, 'viewset_instance': self})
+                else:
+                    serializer_registry[cls_name] = type(cls_name, (AssociationSerializerBase,), {'target_queryset': qs})
+            return serializer_registry[cls_name]
+
         return super().get_serializer_class()
 
 
+def filter_queryset_no_op(qs: QuerySet) -> QuerySet:
+    return qs
+
+
+class AssociationSerializerBase(serializers.Serializer):
+    """Serializer used for associating related objects, where all those related objects are allowed
+
+    It is expected that subclasses will set target_queryset which gives the queryset
+    for the model that will be associated"""
+
+    target_queryset = None
+
+    def get_queryset_on_init(self) -> QuerySet:
+        return self.target_queryset
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.fields['instances'] = serializers.PrimaryKeyRelatedField(queryset=self.get_queryset_on_init(), many=True)
+
+
+class FilteredAssociationSerializer(AssociationSerializerBase):
+    viewset_instance = None
+
+    def get_queryset_on_init(self) -> QuerySet:
+        qs = super().get_queryset_on_init()
+        if self.viewset_instance and hasattr(self.viewset_instance, 'filter_associate_queryset'):
```

Comment:
This callback could still be an issue... and this also leaves a viewset object referenced by objects in the registry which creates garbage collection problems. So it might be better, if possible, to refer to the view inside of the context. As I previously deleted `request = self.context['request']`. We might want to resurrect that to get the view instead of the request. Like this:

https://github.com/ansible/awx/blob/d4bdaad4d83a912d05dfbc833b5aeef4e915403c/awx/conf/serializers.py#L64